File: examples/chapter13/grid2ddiff.m
%%
% We define a function and its two exact partial derivatives. 
u = @(x,y) sin(x.*y-y);
dudx = @(x,y) y.*cos(x.*y-y);
dudy = @(x,y) (x-1).*cos(x.*y-y);

%%
% We use an equispaced grid and second order finite differences as
% implemented by |diffmat2|.
m = 80;  [x,Dx] = diffmat2(m,[0,2*pi]);
n = 60;  [y,Dy] = diffmat2(n,[1,3]);
[U,X,Y] = mtx(u,x,y);

%%
% Here we compare the exact $\partial u/\partial x$ with the
% finite-difference approximation.
subplot(1,2,1)
contourf(X,Y,dudx(X,Y),10)
title('du/dx')     % ignore this line
subplot(1,2,2)
contourf(X,Y,Dx*U,10)
title('approximation')    % ignore this line

%%
% We now do the same for $\partial u/\partial y$,
subplot(1,2,1)
contourf(X,Y,dudy(X,Y),10)
title('du/dy')    % ignore this line
subplot(1,2,2)
contourf(X,Y,U*Dy',10)
title('approximation')    % ignore this lineo

%%
% To the eye there is little difference to be seen, though we expect that
% the results are probably have no more than 2-4 correct digits for these
% discretization sizes. 

File: julia/scripts/chapter13/grid2ddiff.jl

u(x,y) = sin(x*y-y);
dudx(x,y) = y*cos(x*y-y);
dudy(x,y) = (x-1)*cos(x*y-y);

include("../FNC.jl")
m = 80;  x,Dx = FNC.diffmat2(m,[0,2*pi]);
n = 60;  y,Dy = FNC.diffmat2(n,[1,3]);
X,Y = FNC.ndgrid(x,y)
U = u.(X,Y);

plot(X,Y,dudx.(X,Y),match_dimensions=true,levels=10,
    title="du/dx") 

using Plots

contourf(x,y,dudx.(X,Y),match_dimensions=true,levels=10,layout=(1,2),subplot=1,
    title="du/dx")      
contourf!(x,y,Dx*U,levels=10,match_dimensions=true,layout=(1,2),subplot=2,
    title="approximation")      

contourf(x,y,dudy.(X,Y),match_dimensions=true,levels=10,layout=(1,2),subplot=1,
    title="du/dy")      
contourf!(x,y,U*Dy',levels=10,match_dimensions=true,layout=(1,2),subplot=2,
    title="approximation")
